New config structure docs (#96)
feature-sliced/steiger: README.md

> [!NOTE]
> The project is in beta and in active development. Some APIs may change.

> [!NOTE]

> > Version 0.5.0 introduces a new config file format. Please refer to the [Configuration](#configuration) section for more information and migration guide.

Reviewer: issue: the migration guide is quite far down that section

suggestion: let's just put the link here:
```suggestion
> Version 0.5.0 introduced a new config file format. We have a codemod to automatically update your config, see the [migration guide](./MIGRATION_GUIDE.md).
```

Author: Yes, makes sense

steiger ./src --watch
```

## Configuration

Steiger is zero-config! If you don't want to disable certain rules, you can safely skip this section.

> Steiger is configurable via `cosmiconfig`. That means that you can create a `steiger.config.ts` or `steiger.config.js` file in the root of your project to configure the rules. Import `{ defineConfig } from "steiger"` to get autocompletion.

Reviewer: issue: this section is no longer accurate with the new changes

suggestion: let's leave the type safety for a future PR
```suggestion
Steiger is configurable via `cosmiconfig`. That means that you can create a `steiger.config.ts` or `steiger.config.js` file in the root of your project to configure the rules.
```

The config file shape is highly inspired by ESLint's config file, so if you have configured ESLint before, you'll find it easy to configure Steiger.

### Example

Here are some rules on how configuration is processed:

> - Config objects are processed from top to bottom, so if there are multiple config object that match the same file for the same rule, the last one will be applied.

Reviewer: issue: I think it might be slightly unclear what "config objects" are, they might be confused with the entire file.

suggestion:
```suggestion
- Objects in the config array are processed in order, so later objects can override the objects before them.
```

> - You can set options for a rule once. When set, options are applied for the entire file system that is covered by Steiger.

Reviewer: issue: I think we also need to clarify here that it's not just that you can set it once, you should set it once

suggestion:
```suggestion
- If a rule has configurable options, then you can set it once in any of the objects in the array and these options will be applied to the entire codebase, regardless of `files` and `excludes`.
```

Note that this line `...fsd.configs.recommended,` just takes the plugin and the recommended rules configuration (all enabled with "error" severity by default) and puts it into the config array.

```javascript
// ./steiger.config.ts
import fsd from '@feature-sliced/steiger-plugin'
import defineConfig from 'steiger'

export default defineConfig([
  ...fsd.configs.recommended,
  {
    // ignore all mock files for all rules
    ignores: ['**/__mocks__/**'],
  },
  {

>     files: ['**/shared/**'],

Reviewer: suggestion (non-blocking): ideally I think we should encourage people to provide precise paths to layers, otherwise if there's a folder called "shared" somewhere deep within, it will also be targeted
```suggestion
    files: ['./src/shared/**'],
```

npx steiger ./src --watch
```

## Configuration

Steiger is zero-config! If you don't want to disable certain rules, you can safely skip this section.

Steiger is configurable via `cosmiconfig`. That means that you can create a `steiger.config.ts` or `steiger.config.js` file in the root of your project to configure the rules. Import `{ defineConfig } from "steiger"` to get autocompletion.

The config file shape is highly inspired by ESLint's config file, so if you have configured ESLint before, you'll find it easy to configure Steiger.

### Example

> Here are some rules on how configuration is processed:

Reviewer: issue: when I see the header "Example", I expect to see a simple example right away, but in this version there's a big blob of text to go through first :D

suggestion: let's put a small example first, like this one:
```js
// ./steiger.config.js
import { defineConfig } from 'steiger'
import fsd from '@feature-sliced/steiger-plugin'

export default defineConfig([
  ...fsd.configs.recommended,
  {
    // disable the `public-api` rule for files in the Shared layer
    files: ['./src/shared/**'],
    rules: {
      'fsd/public-api': 'off',
    },
  }
])
```

Author: No problem

> - Config objects are processed from top to bottom, so if there are multiple config object that match the same file for the same rule, the last one will be applied.
> - GlobalIgnore objects `{ ignores: ['**/__mocks__/**'] }` are applied to all rules, they are processed first and permanently remove files from the linter's field of view, so you can't reassign severity to them in other later objects.
> - You can set options for a rule once. When set, options are applied for the entire file system that is covered by Steiger.
> 
> Note that this line `...fsd.configs.recommended,` just takes the plugin and the recommended rules configuration (all enabled with "error" severity by default) and puts it into the config array.

Reviewer: suggestion: let's not try to document every single aspect of behaviour. I think this makes instructions too big, which means that people will be less likely to read them.

For instance:
- "Config objects are processed from top to bottom" — I think this is something that they can discover by trying, and most people are probably not even going to encounter that
- "GlobalIgnore objects" — I think this goes a bit too deep into the internal terminology like "GlobalIgnore", "assign severity", "linter's field of view", I don't think that's super relevant. I suggest simplifying it
- "You can set options for a rule once." — at present, none of the rules even accept options, I think, so this is very much irrelevant :D
- About the `fsd.configs.recommended`, I don't think we need to tell people how it works exactly, all they need to know is that it will set up the linter the way we recommend, and they can tweak it further if they want.

I don't even think it should be a list, let's just throw a tip like this after that simple example:
```suggestion
> [!TIP]
> If you want Steiger to ignore certain files, add an object like this to the config array: 
> ```js
> defineConfig([
>   /* … */, 
>   { ignores: ['**/__mocks__/**'] }
> ])
> ```
```

Author: I agree, most of this section is either irrelevant or too complex

```javascript
// ./steiger.config.ts
import fsd from '@feature-sliced/steiger-plugin'

> import defineConfig from 'steiger'

Reviewer: suggestion: let's put this example under a spoiler so that it's there if you want to go further, but doesn't clutter the README for the people who want to speed-run the README:

```
<details>
  <summary>Comprehensive showcase of the config file syntax</summary>
  
```js
...
```

</details>
```

Reviewer: also, issue: the example shows `import defineConfig from 'steiger'`, it should be `import { defineConfig } from 'steiger'`

Author: That sounds completely reasonable to me

  ...fsd.configs.recommended,
  {
    // ignore all mock files for all rules
    ignores: ['**/__mocks__/**'],
  },
  {
    files: ['./src/shared/**'],
    rules: {
      // disable public-api rule for files in /shared folder
      'fsd/public-api': 'off',
    },
  },
  {
    files: ['./src/widgets/**'],
    ignores: ['**/discount-offers/**'],
    rules: {
      // disable no-segmentless-slices rule for all widgets except /discount-offers
      'fsd/no-segmentless-slices': 'off',
    },
  },
])
```

### Glob matching

All globs are matched only against files, folder severities are computed based on the files inside them. The formula is simple: the folder severity is the highest severity of files inside it (from highest to lowest: error, warn, off).

**Glob examples**:

- `./src/shared/**` - matches all files in the `shared` folder and its subfolders
- `./src/shared/*` - matches all files that are direct children of the `shared` folder
- `./src/shared` - based on the fact that globs are matched against files, this one matches only `shared` file (without an extension) inside the `src` folder
- `**/__mocks__/**` - matches all files in all `__mocks__` folders throughout the project

> - `**/*.{test,spec}.{ts,tsx}` - matches all test files in the project

Reviewer: issue: again, I believe this is a more technical description, and I forecast that most people will just disable entire folders, so people are unlikely to run up against a case where one file made the whole folder be `error`.

suggestion: let's just get rid of this entire section

Author: Decided to move it to CONFIG_EXAMPLES.md so far. Let me know if you think it should not be even there

Reviewer: I personally think it shouldn't be even there, but I also don't mind it being there. As long as it's not up front in the README, I'm fine with it
